pkg/common/commands/resetoffset/controller/offsets.go

```diff
@@ -0,0 +1,210 @@
+	offsetMappings := make([]kafkav1alpha1.OffsetMapping, len(partitions))
+
+	// Loop Over The Partitions - Tracking Status
+	commitOffsets := true
+	for index, partition := range partitions {
+
+		// Enhance The Logger With Partition
+		partitionLogger := logger.With(zap.Int32("Partition", partition))
+
+		// Update The Partition's Offset (Process all partitions even on error in order to populate OffsetMappings for CRD)
+		offsetMapping, updateErr := updateOffset(partitionLogger, saramaClient, offsetManager, topicName, partition, offsetTime)
+		if offsetMapping != nil {
+			offsetMappings[index] = *offsetMapping
+		}
+		if updateErr != nil {
+			partitionLogger.Error("Failed to update offset", zap.Error(updateErr))
+			commitOffsets = false
+		} else {
+			partitionLogger.Info("Successfully updated offset", zap.Any("OffsetMapping", offsetMapping))
+		}
+	}
+
+	// If All Partitions Were Updated Successfully Then Commit The New Offsets
+	if commitOffsets {
+		logger.Info("All Offsets updated successfully - performing Commit")
+		offsetManager.Commit()
+		return offsetMappings, nil
+	} else {
+		logger.Error("Failed to update all Offsets - skipping Commit")
+		return offsetMappings, updateOffsetsError
+	}
+}
+
+// updateOffset performs the update of a single Partition's Offset and
+// returns an OffsetMapping representing the old/new state.
+func updateOffset(logger *zap.Logger,
+	saramaClient sarama.Client,
+	offsetManager sarama.OffsetManager,
+	topic string,
+	partition int32,
+	offsetTime int64) (*kafkav1alpha1.OffsetMapping, error) {
+
+	// Create A PartitionOffsetManager For The Specified Partition
+	partitionOffsetManager, err := offsetManager.ManagePartition(topic, partition)
+	defer safeClosePartitionOffsetManager(logger, partitionOffsetManager)
+	if err != nil {
+		logger.Error("Failed to create PartitionOffsetManager", zap.Error(err))
+		return nil, err
+	}
+
+	// Get The Current Offset Of Partition (Accuracy Depends On ConsumerGroup Having Been Stopped)
+	currentOffset, _ := partitionOffsetManager.NextOffset()
+
+	// Get The New Offset Of Partition For Specified Time
+	newOffset, err := saramaClient.GetOffset(topic, partition, offsetTime)
+	if err != nil {
+		logger.Error("Failed to get Partition Offset for Time", zap.Int64("Time", offsetTime), zap.Error(err))
+		return nil, err
+	}
+
+	// Create An OffsetMapping For The Partition
+	offsetMapping := &kafkav1alpha1.OffsetMapping{
+		Partition: partition,
+		OldOffset: currentOffset,
+		NewOffset: newOffset,
+	}
+
+	// Update The Partition's Offset Forward/Back As Needed
+	offsetMetaData := formatOffsetMetaData(offsetTime)
+	if newOffset > currentOffset {
+		partitionOffsetManager.MarkOffset(newOffset, offsetMetaData)
+	} else if newOffset < currentOffset {
+		partitionOffsetManager.ResetOffset(newOffset, offsetMetaData)
+	}
+
+	// Check For Any Errors From The PartitionOffsetManager
+	select {
+	case err = <-partitionOffsetManager.Errors():
+	case <-time.After(1 * time.Second):
+	}
+
```

Comment:
Yeah, I agree fully and am not very happy with this current implementation.  Obviously anytime you're putting in a "sleep" or "wait" duration it's not good.  

This IS how the Sarama PartitionOffsetManager API is setup to work.  The `MarkOffset()` and `ResetOffset()` functions do NOT return anything, and you are intended to instead monitor the `Errors()` Channel.  The parent OffsetManager does not expose any error handling (say for `Commit()`)

```
	// Errors returns a read channel of errors that occur during offset management, if
	// enabled. By default, errors are logged and not returned over this channel. If
	// you want to implement any custom error handling, set your config's
	// Consumer.Return.Errors setting to true, and read from this channel.
	Errors() <-chan *ConsumerError
```

I think the correct way to improve this is to re-arrange the lifecycle handling of PartitionOffsetManager so that we wait for the `Errors()` channel to close AFTER closing the PartitionOffsetManager.  This is possible but complicates the implementation since we can't just "defer" the closing, and instead have to handle it in all exit cases.

I think what I have above was a temporary/lazy solution while I was busy working on the rest of the controller.  I can/will look into fixing that as described above but would prefer to do it outside (immediately after) this PR if that's acceptable.  The Controller is incomplete and not actively used anywhere so developing it incrementally should hopefully be acceptable?

Reply:
I had some time this afternoon to refactor that logic so it should be better now.  There is still a `time.After()` usage but it is no longer a vague guess at how long it might take the PartitionOffsetManager to write a ConsumerError to the Error channel.  Instead it is a large-enough window for the PartitionOffsetManager to close the Error channel once it is closing itself.  We're no longer `defer()`'ing the PartitionOffsetManager `close()` after we try to process the channel, but are instead closing first which should close the error channel, and are then draining the error channel.  This better aligns with the Sarama documentation on `AsyncClose()` (called by `Close()`)...
```
	// AsyncClose initiates a shutdown of the PartitionOffsetManager. This method will
	// return immediately, after which you should wait until the 'errors' channel has
	// been drained and closed. It is required to call this function, or Close before
	// a consumer object passes out of scope, as it will otherwise leak memory. You
	// must call this before calling Close on the underlying client.
	AsyncClose()
```

Reply:
Just checked right now. More than having the `drainPartitionOffsetManagerErrors`, I would love to have an internal API that hides completely that weirdness, something like:

```
MarkOffset(partitionOffsetManager, newOffset, offsetMetaData) error
```

So when I read this code, I don't need to reason about the bad Sarama API. WDYT?

Reply:
I don't think it's that easy.   The function would have to be "MarkOffsetAndCloseParttitionOffsetManager()" because it's important to close the PartitionOffsetManager prior to draining.  Then we're combining the "end" of the process (marking offset, closing partition manager, and draining error channel) all into one big step and we have to repeat that for ResetOffset.  So it doesn't end up feeling very clean to me.

I think this is just how the Sarama library works and we're dealing with it in a good way.  FWIW... This is the same way the Sarama ConsumerGroup Error handling works (separate async channel) and we haven't tried to do anything there to "hide" the nature of the library.

I would like to keep it this way for now - I don't think it's too confusing (not any worse than some buried/wrapped solution that tries to hide what the Sarama library is).   If it's a problem for some reason going forward we can look at changing it then?

Reply:
**TLDR...**
Well, I take it all back - haha ; ).    Seems I wasn't doing a very good job of handling the Sarama errors.  Don't get me wrong, the Sarama API is still a wonky mess in this area but... my understanding of how it worked was lacking.  I have re-written most of the `offsets.go` file which should probably be reviewed as a new file (easier that way).

**The details...**
My confusion was based on the expectation that the Parent / Child nature of the OffsetManager & PartitionOffsetManager was "logical".  I was creating and Closing() the PartitionOffsetManager in the sub-loop, and was expecting the `MarkOffset()` and `ResetOffset()` calls to possibly generate errors on the `Errors()` channel.  After re-reading the docs for those functions and a fair amount of their source it seems that it's not that straightforward.

The OffsetManager and the PartitionOffsetManager are tightly intertwined and there are expectations about their usage.  The OffsetManager must be closed after the PartitionOffsetManagers, but the PartitionOffsetManager `Close()` will block on returning errors until the OffsetManager is closed (when autocommit is disabled).  Further the expectation is that you drain the PartitionOffsetManager `Errors()` channel AFTER closing the OffsetManager, etc, etc..

The core logic is mosty the same, just the order of things has changed.

Thanks in advance for re-reviewing!